<a href="https://colab.research.google.com/github/MazRadwan/data_science/blob/main/1_Introduction_to_Pandas_DataFrames.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python and Data Analysis 1 - Introduction to Pandas DataFrames

**Goal:** The goal of this project is to become comfortable working with data in Pandas.

**Description:** This project will cover the basics of Pandas: importing, manipulating, and filtering data in Pandas DataFrames. Becoming fluent takes practice, but is necessary when building larger data analysis projects. After this project, you should feel comfortable navigating any dataset using Pandas.

## 1A: Data Basics

### Types of Data

At the most fundemental level, there are three kinds of data:
 - *Qualitative/Categorical*: Country, Industry, Faculty...
 - *Quantitative/Numerical*: Height, Rank, Price...
 - *Identifying*: Stock Ticker, Card Number, Product ID

Pandas DataFrames can store different data types, but the basic ones we will focus on here are:
 - *Strings* (`object`): Text
 - *Integers* (`int64`): Numbers without decimals
 - *Floats* (`float64`): Numbers with decimals
 - *Booleans* (`bool`): Values that are either True or False
 - *DateTimes* (`datetime64`): Values that store a specific date and time

### Importing and Displaying Data

We can import data from CSV (comma separated value) files with `read_csv`. The following stock dataset is sourced from https://www.macrotrends.net/.

In [ ]:
import pandas as pd
df = pd.read_csv('MSFT.csv') # Import the CSV

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

We can then print the entire DataFrame.

In [ ]:
# df.tail()

Print the first five rows with `head`.

In [ ]:
print(df.head())

         date    open    high     low   close      volume
0  1986-03-13  0.0885  0.1016  0.0885  0.0972  1031788800
1  1986-03-14  0.0972  0.1024  0.0972  0.1007   308160000
2  1986-03-17  0.1007  0.1033  0.1007  0.1024   133171200
3  1986-03-18  0.1024  0.1033  0.0990  0.0998    67766400
4  1986-03-19  0.0998  0.1007  0.0972  0.0981    47894400


Print the last five rows with `tail`.

In [ ]:
print(df.tail())

            date     open    high       low   close    volume
8629  2020-06-08  185.940  188.55  184.4400  188.36  33123035
8630  2020-06-09  188.000  190.70  187.2605  189.80  29783916
8631  2020-06-10  191.125  198.52  191.0100  196.84  43568260
8632  2020-06-11  193.130  195.76  186.0700  186.27  52645278
8633  2020-06-12  190.540  191.72  185.1800  187.74  43373587


Use `dtype` to determine the type of data.

In [ ]:
df['close'].dtype

dtype('float64')

In [ ]:
df['close'].dtype # Find the type of data in the 'close' column

dtype('float64')

**Challenge**: Guess what the data type of the 'volume' column is, then write code to display it.

In [ ]:
df['volume'].dtype

dtype('int64')

### DataFrame Elements

DataFrames have three key elements:
 - *Data*: The content of the DataFrame
 - *Columns*: Horizontal headings (`date`, `open`, `high`, `low`, `close`, `volume`) uniquelely identifying a column
 - *Index*: Vertical values uniquely identifying a row (the numbers on the leftmost side)


Note: The *position* of a column or index is the order in which it appears, starting from 0. The column positions are numbered 0, 1, 2, 3, 4, 5, in order of appearance. In this DataFrame, the index is the same as the index position for all rows (index at position 0 has value 0).

## 1B: Slicing and Dicing Data

When analyzing data, we need to be able to manipulate DataFrames to select the exact subset of data that we want. Here, we will cover a few key techniques.

### Column Select

If we want to get data from a particular column, we can use the syntax `df-name['column-name']`. This returns a Pandas *Series*. For example, if we want to get the `date` column from the DataFrame called `df`, we can do:

In [ ]:
dates = df['date']
dates = df['volume']
print(dates)

0       1031788800
1        308160000
2        133171200
3         67766400
4         47894400
           ...    
8629      33123035
8630      29783916
8631      43568260
8632      52645278
8633      43373587
Name: volume, Length: 8634, dtype: int64


To select multiple columns, we use `df-name[['col1', 'col2', ...]]`. Instead of passing a single column, we pass a list of columns. Instead of returning a Series, Pandas will return a DataFrame. This also allows us to rearrange the order of columns.

In [ ]:
multiple_cols = df[['volume', 'close', 'date']]
# multiple_cols = df[['volume', 'close', 'date ']]
print(multiple_cols)
multiple_cols

          volume     close        date
0     1031788800    0.0972  1986-03-13
1      308160000    0.1007  1986-03-14
2      133171200    0.1024  1986-03-17
3       67766400    0.0998  1986-03-18
4       47894400    0.0981  1986-03-19
...          ...       ...         ...
8629    33123035  188.3600  2020-06-08
8630    29783916  189.8000  2020-06-09
8631    43568260  196.8400  2020-06-10
8632    52645278  186.2700  2020-06-11
8633    43373587  187.7400  2020-06-12

[8634 rows x 3 columns]


volume     close        date
0     1031788800    0.0972  1986-03-13
1      308160000    0.1007  1986-03-14
2      133171200    0.1024  1986-03-17
3       67766400    0.0998  1986-03-18
4       47894400    0.0981  1986-03-19
...          ...       ...         ...
8629    33123035  188.3600  2020-06-08
8630    29783916  189.8000  2020-06-09
8631    43568260  196.8400  2020-06-10
8632    52645278  186.2700  2020-06-11
8633    43373587  187.7400  2020-06-12

[8634 rows x 3 columns]

**Challenge**: Print out the 'high' and 'low' columns.

In [ ]:
print(df[['high', 'low']])

          high       low
0       0.1016    0.0885
1       0.1024    0.0972
2       0.1033    0.1007
3       0.1033    0.0990
4       0.1007    0.0972
...        ...       ...
8629  188.5500  184.4400
8630  190.7000  187.2605
8631  198.5200  191.0100
8632  195.7600  186.0700
8633  191.7200  185.1800

[8634 rows x 2 columns]


### Row Select

Pandas provides two ways to select particular rows from a DataFrame.

#### Row Select by Value

We use the row's index value to select the row, following the syntax `df_name.loc[index_value]`. This will return a Series. Here, the index value is the same as its position (the index value at position 0 is 0), but this is not always the case. The index value can be a non-integer, but its position will always be an integer from 0 to the number of rows minus one.

In [ ]:
df.loc[0]

df.loc[[0,1,2,3,4 ]]


date    open    high     low   close      volume
0  1986-03-13  0.0885  0.1016  0.0885  0.0972  1031788800
1  1986-03-14  0.0972  0.1024  0.0972  0.1007   308160000
2  1986-03-17  0.1007  0.1033  0.1007  0.1024   133171200
3  1986-03-18  0.1024  0.1033  0.0990  0.0998    67766400
4  1986-03-19  0.0998  0.1007  0.0972  0.0981    47894400

In [ ]:
first_row = df.loc[0]
print(first_row)

date      1986-03-13
open          0.0885
high          0.1016
low           0.0885
close         0.0972
volume    1031788800
Name: 0, dtype: object


For multiple rows, we can pass a list of rows, or a range using `df_name.loc[[values_list]]` or `df_name.loc[start_value:end_value]`.

In [ ]:
first_five_rows = df.loc[[0,1,2,3,4]] # Access the rows with index 0,1,2,3,4
print(first_five_rows)

first_five_rows_range = df.loc[0:4] # Access the rows starting with index 0 up to and including index 4
print(first_five_rows_range)

         date    open    high     low   close      volume
0  1986-03-13  0.0885  0.1016  0.0885  0.0972  1031788800
1  1986-03-14  0.0972  0.1024  0.0972  0.1007   308160000
2  1986-03-17  0.1007  0.1033  0.1007  0.1024   133171200
3  1986-03-18  0.1024  0.1033  0.0990  0.0998    67766400
4  1986-03-19  0.0998  0.1007  0.0972  0.0981    47894400
         date    open    high     low   close      volume
0  1986-03-13  0.0885  0.1016  0.0885  0.0972  1031788800
1  1986-03-14  0.0972  0.1024  0.0972  0.1007   308160000
2  1986-03-17  0.1007  0.1033  0.1007  0.1024   133171200
3  1986-03-18  0.1024  0.1033  0.0990  0.0998    67766400
4  1986-03-19  0.0998  0.1007  0.0972  0.0981    47894400


**Challenge**: Print out rows with index values 1000 to 2000, then print out rows with index values 1000, 1500, and 2000.

In [ ]:
print(df.loc[1000:2000])
print(df.loc[[1000, 1500, 2000]])

            date    open    high     low   close     volume
1000  1990-02-26  0.6771  0.6944  0.6753  0.6892   69219216
1001  1990-02-27  0.6910  0.6910  0.6806  0.6806   62019216
1002  1990-02-28  0.6806  0.6927  0.6806  0.6858   39312000
1003  1990-03-01  0.6840  0.7066  0.6840  0.6944   76867200
1004  1990-03-02  0.6962  0.7248  0.6953  0.7205  100208016
...          ...     ...     ...     ...     ...        ...
1996  1994-02-01  2.6563  2.6797  2.6406  2.6600   44003200
1997  1994-02-02  2.6484  2.6719  2.6250  2.6250   40924800
1998  1994-02-03  2.6328  2.6406  2.6094  2.6406   31792000
1999  1994-02-04  2.6250  2.6406  2.5391  2.5391   78217600
2000  1994-02-07  2.4922  2.5391  2.4844  2.5234  118851200

[1001 rows x 6 columns]
            date    open    high     low   close     volume
1000  1990-02-26  0.6771  0.6944  0.6753  0.6892   69219216
1500  1992-02-18  2.4844  2.5469  2.4583  2.4635   73692816
2000  1994-02-07  2.4922  2.5391  2.4844  2.5234  118851200


#### Row Select by Position

Instead of using the value of the index, we can also use its position. This is useful if our index values are not common-sense integers. If our index was a string, but we wanted to select the first 20 rows of data, `loc` would not work unless we knew the first 20 values in our index. With the exception of using positions instead of values, `iloc` is quite similar to `loc`. We use the syntax `df_name.iloc[index_position]` for a single row and `df_name.iloc[[positions_list]]` or `df_name.iloc[start_position:end_position]` for multiple rows.

In [ ]:
row_at_position_4 = df.iloc[4] # Access row with position 4
print(row_at_position_4)

last_five_rows = df.iloc[[-5,-4,-3,-2,-1]] # Access rows with positions -5,-4,-3,-2,-1
print(last_five_rows)

last_five_rows_range = df.iloc[-5:] # Access rows with positions -5 to the end
print(last_five_rows_range)

date      1986-03-19
open          0.0998
high          0.1007
low           0.0972
close         0.0981
volume      47894400
Name: 4, dtype: object
            date     open    high       low   close    volume
8629  2020-06-08  185.940  188.55  184.4400  188.36  33123035
8630  2020-06-09  188.000  190.70  187.2605  189.80  29783916
8631  2020-06-10  191.125  198.52  191.0100  196.84  43568260
8632  2020-06-11  193.130  195.76  186.0700  186.27  52645278
8633  2020-06-12  190.540  191.72  185.1800  187.74  43373587
            date     open    high       low   close    volume
8629  2020-06-08  185.940  188.55  184.4400  188.36  33123035
8630  2020-06-09  188.000  190.70  187.2605  189.80  29783916
8631  2020-06-10  191.125  198.52  191.0100  196.84  43568260
8632  2020-06-11  193.130  195.76  186.0700  186.27  52645278
8633  2020-06-12  190.540  191.72  185.1800  187.74  43373587


**Challenge**: Select the last 2000 rows in the dataframe **except for the last row**. Hint: the last row always has position -1.

In [ ]:
df.iloc[[-5,-4,-3,-2,-1]]

# specify column values instead of the row values
df.iloc[:5,[4, 5]]

print()

df.loc[:5, ['close', 'volume']]

close      volume
0  0.0972  1031788800
1  0.1007   308160000
2  0.1024   133171200
3  0.0998    67766400
4  0.0981    47894400
5  0.0955    58435200

In [ ]:
# Practice: try seledcting the high and low prices for a few
# rows in the middle, and first and last rows using loc and iloc

df.loc[[1000, 1500, 2000], ['high', 'low']]

df.iloc[[1000, 1500, 2000], [2, 3]]

df.loc[[0, 5000, 8633], ['high', 'low']]

df.iloc[[0, 5000, -1], [2, 3]]





high       low
0       0.1016    0.0885
5000   27.1300   26.9100
8633  191.7200  185.1800

In [ ]:
print(df.iloc[-2000:-1])

            date     open    high       low    close    volume
6634  2012-07-02   30.620   30.62   30.2100   30.560  30589100
6635  2012-07-03   30.230   30.76   30.1700   30.760  20938100
6636  2012-07-05   30.590   30.78   30.3750   30.700  28801900
6637  2012-07-06   30.610   30.70   29.9500   30.185  38294800
6638  2012-07-09   30.120   30.23   29.7800   30.000  30680800
...          ...      ...     ...       ...      ...       ...
8628  2020-06-05  182.620  187.73  182.0100  187.200  39893643
8629  2020-06-08  185.940  188.55  184.4400  188.360  33123035
8630  2020-06-09  188.000  190.70  187.2605  189.800  29783916
8631  2020-06-10  191.125  198.52  191.0100  196.840  43568260
8632  2020-06-11  193.130  195.76  186.0700  186.270  52645278

[1999 rows x 6 columns]


### Subsection Select

To select a subsection of rows and columns at the same time, we can use either `df_name.loc[row_values, column_values]` or `df_name.iloc[row_positions, column_positions]`. The key difference is `loc` takes the value itself, whereas `iloc` takes the position.

In [ ]:
first_5_close_and_vol = df.loc[[0,1,2,3,4], ['close', 'volume']] # Access the 'close' and 'volume' columns for the rows with index values 0,1,2,3,4
print(first_5_close_and_vol)

last_2_date_close_and_volume = df.iloc[-2:, [0,4,5]] # Access the 'date' and 'close' and 'volume' columns for rows with index position -2 to the end
print(last_2_date_close_and_volume)

    close      volume
0  0.0972  1031788800
1  0.1007   308160000
2  0.1024   133171200
3  0.0998    67766400
4  0.0981    47894400
            date   close    volume
8632  2020-06-11  186.27  52645278
8633  2020-06-12  187.74  43373587


**Challenge**: Select and display the `high` and `low` prices for the first row, last row, and row 5000 using both `loc` and `iloc`. When you use `iloc`, try to ensure it accesses the last row regardless of the number of rows in the dataframe.

In [ ]:
print(df.loc[[0, 5000, 8633], ['high', 'low']])
print(df.iloc[[0, 5000, -1], [2, 3]])

          high       low
0       0.1016    0.0885
5000   27.1300   26.9100
8633  191.7200  185.1800
          high       low
0       0.1016    0.0885
5000   27.1300   26.9100
8633  191.7200  185.1800


## 1C: Filtering Data

In [ ]:
# write a condition in the brackets

df[df['volume'] > 355000000]

date     open     high      low    close      volume
0     1986-03-13   0.0885   0.1016   0.0885   0.0972  1031788800
754   1989-03-07   0.4149   0.4167   0.3698   0.3715   788688000
2081  1994-06-06   3.2894   3.4144   3.2344   3.4063   764504000
5078  2006-04-28  24.2600  24.5000  24.0000  24.1500   591052200

Perhaps more interesting than manually selecting data, *filtering* allows us to access a subset of rows based on a condition or set of conditions. For example, if we were only interested in rows where the high price was greater than 1, we would filter out all other rows. To do this, we use the syntax `df_name[condition]`.

In [ ]:
filtered = df[df['high'] > 1] # Select only the rows where high > 1
print(filtered)

            date      open      high       low     close     volume
1059  1990-05-21    0.9549    1.0035    0.9549    0.9757  159593616
1060  1990-05-22    0.9861    1.0035    0.9583    0.9931  142435224
1061  1990-05-23    1.0000    1.0556    0.9982    1.0538  141180768
1062  1990-05-24    1.0660    1.0938    1.0556    1.0625  146649600
1063  1990-05-25    1.0313    1.0590    1.0278    1.0313  104688000
...          ...       ...       ...       ...       ...        ...
8629  2020-06-08  185.9400  188.5500  184.4400  188.3600   33123035
8630  2020-06-09  188.0000  190.7000  187.2605  189.8000   29783916
8631  2020-06-10  191.1250  198.5200  191.0100  196.8400   43568260
8632  2020-06-11  193.1300  195.7600  186.0700  186.2700   52645278
8633  2020-06-12  190.5400  191.7200  185.1800  187.7400   43373587

[7488 rows x 6 columns]


We can combine more than one condition using the bitwise comparators `&` (and) and `|` (or). If we wanted the rows where `volume` was larger than 35 Million and `close` was greater than 184, we could combine the two conditions with `&`.

In [ ]:
filtered = df[(df['volume'] > 35000000) & (df['close'] > 100)]
print(filtered)

            date     open    high      low   close    volume
8132  2018-06-15  101.510  101.53  100.070  100.13  65738585
8137  2018-06-22  100.410  100.77   99.625  100.41  38923105
8155  2018-07-19  104.930  105.31  103.890  104.40  40171646
8156  2018-07-20  108.080  108.20  106.080  106.27  56038827
8161  2018-07-27  110.180  110.18  106.140  107.68  37005292
...          ...      ...     ...      ...     ...       ...
8623  2020-05-29  182.730  184.27  180.410  183.25  42146720
8628  2020-06-05  182.620  187.73  182.010  187.20  39893643
8631  2020-06-10  191.125  198.52  191.010  196.84  43568260
8632  2020-06-11  193.130  195.76  186.070  186.27  52645278
8633  2020-06-12  190.540  191.72  185.180  187.74  43373587

[129 rows x 6 columns]


Notice that the number of rows decreases the more conditions we combine. This makes sense, because as we combine conditions, more rows fail to meet the criteria and are filtered out, resulting in a smaller DataFrame.

**Challenge**: Try selecting and displaying only the rows where the difference between the open and close price is greater than 5. Note: you can get the difference between two columns by subtracting them.

In [ ]:
# try selecting and displayin only the rows where the difference between the open and close
# price is greater than 5

df[df['open'] - df['close'] > 5]


date      open      high       low   close    volume
8076  2018-03-27   94.9400   95.1390   88.5100   89.47  56569011
8213  2018-10-10  111.2400  111.5000  105.7900  106.16  61376315
8223  2018-10-24  108.4100  108.4900  101.5901  102.32  63897759
8548  2020-02-11  190.6500  190.7000  183.5000  184.44  53159906
8557  2020-02-25  174.2000  174.8400  167.6462  168.07  68073295
8559  2020-02-27  163.3200  167.0300  157.9800  158.18  93174906
8562  2020-03-03  173.8000  175.0000  162.2600  164.51  71677019
8569  2020-03-12  145.3000  153.4700  138.5800  139.06  93226366
8575  2020-03-20  146.0000  147.1000  135.8600  137.35  84866215
8587  2020-04-07  169.5923  170.0000  163.2600  163.49  62415153
8596  2020-04-21  173.5000  173.6699  166.1102  167.82  56203749
8601  2020-04-28  175.4556  175.6700  169.3900  169.81  33687238
8632  2020-06-11  193.1300  195.7600  186.0700  186.27  52645278